test/distributed/org/apache/cassandra/distributed/action/GossipHelper.java

```diff
@@ -0,0 +1,455 @@
+        }
+    }
+
+    private static byte[] toBytes(EndpointState epState)
+    {
+        try (DataOutputBuffer out = new DataOutputBuffer(1024))
+        {
+            EndpointState.serializer.serialize(epState, out, MessagingService.current_version);
+            return out.toByteArray();
+        }
+        catch (IOException e)
+        {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static EndpointState fromBytes(byte[] bytes)
+    {
+        try (DataInputBuffer in = new DataInputBuffer(bytes))
+        {
+            return EndpointState.serializer.deserialize(in, MessagingService.current_version);
+        }
+        catch (Throwable t)
+        {
+            throw new RuntimeException(t);
+        }
+    }
+
+    private static class PullSchemaFrom implements InstanceAction
+    {
+        final InetSocketAddress pullFrom;
+
+        public PullSchemaFrom(IInvokableInstance pullFrom)
+        {
+            this.pullFrom = pullFrom.broadcastAddress();;
+        }
+
+        public void accept(IInvokableInstance pullTo)
+        {
+            pullTo.acceptsOnInstance((InetSocketAddress pullFrom) -> {
+                InetAddressAndPort endpoint = toCassandraInetAddressAndPort(pullFrom);
+                EndpointState state = Gossiper.instance.getEndpointStateForEndpoint(endpoint);
+                MigrationManager.scheduleSchemaPull(endpoint, state);
+                MigrationManager.waitUntilReadyForBootstrap();
+            }).accept(pullFrom);
+        }
+    }
+
+    private static class BootstrapAction implements InstanceAction, Serializable
+    {
+        private final boolean joinRing;
+        private final Duration waitForBootstrap;
+        private final Duration waitForSchema;
+
+        public BootstrapAction()
+        {
+            this(true, Duration.ofMinutes(10), Duration.ofSeconds(10));
+        }
+
+        public BootstrapAction(boolean joinRing, Duration waitForBootstrap, Duration waitForSchema)
+        {
+            this.joinRing = joinRing;
+            this.waitForBootstrap = waitForBootstrap;
+            this.waitForSchema = waitForSchema;
+        }
+
+        public void accept(IInvokableInstance instance)
+        {
+            instance.appliesOnInstance((String partitionerString, String tokenString) -> {
+                IPartitioner partitioner = FBUtilities.newPartitioner(partitionerString);
+                List<Token> tokens = Collections.singletonList(partitioner.getTokenFactory().fromString(tokenString));
+                try
+                {
+                    assert StorageService.instance.prepareForBootstrap(waitForSchema.toMillis()).size() == 0;
+                    assert StorageService.instance.bootstrap(tokens, waitForBootstrap.toMillis());
+                    StorageService.instance.setUpDistributedSystemKeyspaces();
+                    if (joinRing)
+                        StorageService.instance.finishJoiningRing(true, tokens);
```

Comment:
for me:

The following isn't needed as line 258 blocks replace

```
// remove the existing info about the replaced node.
                if (!current.isEmpty())
                {
                    Gossiper.runInGossipStageBlocking(() -> {
                        for (InetAddressAndPort existing : current)
                            Gossiper.instance.replacedEndpoint(existing);
                    });
                }
```

If we allow replace later, we would need that logic as well.